### PR TITLE
Do not trigger layout when querying breakpoint (getBreakpoint)

### PR DESCRIPTION
**Pull request by robertberry-zz**
`window.getComputedStyle` triggers layout. A whole lot of different things use `getBreakpoint`, so this should give us a performance boost.

### CODE REVIEW

File: static/src/javascripts/projects/common/utils/detect.js [hunk 1]
```diff
@@ -246,15 +247,38 @@ define([
...
-                    : document.getElementsByTagName('head')[0].currentStyle.fontFamily
-            ).replace(/['",]/g, '') || 'mobile',
-            index;
+    function getDocumentWidth() {
+        var d = document.documentElement;
+        return Math.max(
+            d.clientWidth,
+            d.offsetWidth,
+            d.scrollWidth
+        );
```
Comment (stephanfowler) [inline static/src/javascripts/projects/common/utils/detect.js]: http://fiddle.jshell.net/L6uPV/show/4/

**Comment by stephanfowler** [comment]
:+1: yep

**Comment by kaelig** [comment]
Good catch, but in recent browsers it doesn't seem to trigger layout.

I've ran

```js
setInterval(function() { console.log(window.getComputedStyle(document.body, ':after').getPropertyValue('content')); }, 500);
```

And there is no trace of layout events in the timeline.

Did you get different results?

**Comment by robertberry** [comment]
@kaelig I was debugging in Chrome on my Nexus 5, and saw a layout (~200ms) triggered by .getComputedStyle here. So it does seem to still happen at least on the most recent Android Chrome.

**Comment by kaelig** [comment]
Oh, interesting! 200ms is huge. The layout triggered I've seen in the timeline all came from ads (it's in the latest Chrome so that may not apply to all Chromes).

At the FT I'm reading the value from the `<head>`: 

```scss
window.getComputedStyle(document.head, ':after').getPropertyValue('content')
```

I wonder if that triggers any layout as well (it should at least be less expensive)…

**Comment by kaelig** [comment]
/ping @paullewis who might have an actual explanation on these results.

**Comment by robertberry** [comment]
The layout will be any batched writes, though, so it's not necessarily `getComputedStyle` that's at fault there. Also I was attached to a debugger on Chrome, so that's probably slowing everything down a lot ... 

But yes, I think ads are our biggest pain point at the moment. :disappointed:

**Comment by paullewis** [comment]
Right, it's pretty much that if you called `getComputedStyle()` any pending styles will have to be flushed, and then layout run, so you may or may not see it, depending on where you call it in the lifecycle.

Usual vibe should apply: do style reads _then_ writes to avoid a Forced Sync Layout, which in this case probably means querying the last frame's computed style values and storing them in some global (if you're not already) at the start of the frame of interest.

Thanks for the ping, this was funsies.

**Comment by kaelig** [comment]
@paullewis: Thanks for your wisdom!

It reminds me that one of my colleagues and a (now) Mozillian worked on a library to help batching read/writes as you suggest: https://github.com/wilsonpage/fastdom. I believe it's being used on the FT Web App.

**Comment by paullewis** [comment]
:+1: batch those suckers.

**Comment by robertberry** [comment]
Thanks @paullewis. @kaelig We've started using `fastdom` here, too.

**Pull request merged**